addons/pos_loyalty/models/loyalty_reward.py

```diff
@@ -11,3 +11,20 @@ def _get_discount_product_values(self):
         for vals in res:
             vals.update({'taxes_id': False})
         return res
+
+    def _should_compute_all_discount_product(self, reward):
+        parent_res = super()._should_compute_all_discount_product(reward)
+        if parent_res:
+            return parent_res
+        if 'pos_loading' not in self.env.context:
+            return False  # This is not PoS context
+
+        reward_discount_product_domain = reward._get_discount_product_domain()
+        if not reward_discount_product_domain:
+            return False
+        pos_product_product_loaded_fields = set(
+            self.env['pos.session']._get_pos_product_product_fields()) | {'id'}
+        for leaf in reward_discount_product_domain:
+            if len(leaf) == 3 and leaf[0] not in pos_product_product_loaded_fields:
+                return True
```

Comment:
add a logger warning, store in the context the reward that were forced evaluated